core/src/main/java/org/infinispan/stream/CacheCollectors.java

```diff
@@ -0,0 +1,106 @@
+package org.infinispan.stream;
+
+import org.infinispan.commons.marshall.Externalizer;
+import org.infinispan.commons.marshall.SerializeWith;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+/**
+ * Helper class designed to be used to create a serializable Collector for use with
+ * {@link org.infinispan.CacheStream#collect(Collector)} from a supplier of a collector.  The problem is that the
+ * standard {@link java.util.stream.Collectors} class doesn't provide Serializable Collectors and no way to extend
+ * their functionality, so this class is used instead.
```

Comment:
Can the user still pass a custom non-serializable collector to the `collect()` operation, to be processed on the originator node? Also, I think you should override `collect()` in `CacheStream` and link to this.